### PR TITLE
[SG-1051] Names renewal

### PR DESCRIPTION
**Pull request by tasiov**
# Changes

[Document with testing notes](https://www.notion.so/publicawesome/Names-Renewal-Fix-2d7f49c0be5a41758ca6d87c9972bc13?pvs=4)

The purpose of these changes is to implement a working version of Names renewal. Below are some details.

## Names Marketplace Sudo Params

The following fields have been added to the Names Marketplace Sudo Params object.
```Rust
    /// The maximum number of renewals that can be processed in each block
    pub max_renewals_per_block: u32,
    /// The number of bids to query to when searching for the highest bid
    pub valid_bid_query_limit: u32,
    /// The number of seconds before the current block time that a
    /// bid must have been created in order to be considered valid
    pub valid_bid_seconds_delta: u64,
    /// The percentage of the winning bid that must be paid to renew a name
    pub renewal_bid_percentage: Decimal,
    /// The address with permission to invoke process_renewals
    pub operator: String,
```

## Names Marketplace Messages

The following messages were added to the Names Marketplace:
* ExecuteMsg::MigrateBids
    * Only invokable by the operator address
    * Used to migrate bids from the old index to the new index, which has the token_id prefixes
* ExecuteMsg::Renew
    * Invokable within a specific time window of the renewal time.
    * Used to fully renew a name
* ExecuteMsg::ProcessRenewals
    * Only invokable by the operator address
    * Used to process a set of names that are set to renew
* QueryMsg::AsksByRenewTime
    * Used to query a set of asks by there renew time. Good for checking which will renew soon.
* QueryMsg::AskRenewPrice
    * Used to query a the renew price of an Ask. Can change over time with bids.
* QueryMsg::LegacyBids
    * Used to query the legacy bid index, to check if there are bids that need migrating
* SudoMsg:EndBlock
    * Used to systematically process name renewals

### CODE REVIEW

File: contracts/marketplace/src/state.rs
Comment (jason-c-child) [inline contracts/marketplace/src/state.rs:1]: was the intent to include the `renewal_price` in the `Ask` query response (as mentioned here https://discord.com/channels/755548171941445642/1170386685218000906/1187804552804716644)?

File: contracts/marketplace/src/execute.rs [hunk 1]
```diff
@@ -382,6 +399,61 @@ pub fn execute_accept_bid(
...
+
+    let sudo_params = SUDO_PARAMS.load(deps.storage)?;
+
+    ensure_eq!(
+        info.sender,
+        sudo_params.operator,
+        ContractError::Unauthorized {}
+    );
+
+    let old_bids = legacy_bids()
```
Comment (shanev) [inline contracts/marketplace/src/execute.rs:419]: Is it possible for this iteration to go over the block gas limit?

File: contracts/marketplace/src/execute.rs [hunk 2]
```diff
@@ -382,6 +399,61 @@ pub fn execute_accept_bid(
...
+            let msg = BankMsg::Send {
+                to_address: existing_bid.bidder.to_string(),
+                amount: vec![coin(existing_bid.amount.u128(), NATIVE_DENOM)],
+            };
+            response = response.add_message(msg);
+        } else {
+            store_bid(deps.storage, &bid)?;
+        }
+
+        event = event.add_attribute("bid_key", format!("{}-{}", bid.token_id, bid.bidder));
```
Comment (shanev) [inline contracts/marketplace/src/execute.rs:449]: Any more attributes needed for indexing?

**Review by shanev — APPROVED** [review]
Just a couple comments.

File: contracts/marketplace/src/execute.rs (hunk 1 above)
Comment (tasiov) [reply contracts/marketplace/src/execute.rs:419]: the number of bids to process is an arg that is passed in, so would just need to lower that value if the transactions go over the gas limit

**Review by shanev — APPROVED** [review]

**Pull request merged**